---
title: Causal mediation analysis of PTSD GWAS statistics
author: Yongjin Park
theme: jekyll-theme-minimal
date: "`r Sys.Date()`"
---


```{r include = FALSE}

source('../Util.R')
source('../Util-Figure.R')
library(dplyr)
library(readr)
library(ggplot2)
library(ggrepel)
library(pander)

all.ea.file <- '../result/20180602/obs/gene-pgc_ptsd_all_ea.txt.gz'

ind.types <- c('all', 'civ', 'mil')

data.dir <- '../result/20180602/obs/'

tab1 <- (data.dir %&&% 'gene-pgc_ptsd_' %&&% ind.types %&&% '_ea.txt.gz') %>%
    lapply(FUN=read_tsv) %>% bind_rows() %>%
    mutate(rep = '20180602')

data.dir <- '../result/20180605/obs/'
tab2 <- (data.dir %&&% 'gene-pgc_ptsd_' %&&% ind.types %&&% '_ea.txt.gz') %>%
    lapply(FUN=read_tsv) %>% bind_rows() %>%
    mutate(rep = '20180605')

change.eqtl.name <- function(tab) {
    .eqtls <- c('rosmap-mult', 'mayo-mult', 'geuvadis-mult')
    .eqtls.name <- c('DLPFC', 'TCX', 'LCL')
    tab <- tab %>% mutate(data = factor(data, .eqtls, .eqtls.name))
}

change.gwas.name <- function(tab) {
    .gwas <- c('pgc_ptsd_all_ea', 'pgc_ptsd_mil_ea', 'pgc_ptsd_civ_ea')
    .gwas.name <- c('All', 'Mil', 'Civ')
    tab <- tab %>% mutate(gwas = factor(gwas, .gwas, .gwas.name))
}

merge.stat.tabs <- function(tab1, tab2) {

    .cols <- c('med.id', 'hgnc', 'factor', 'data', 'gwas', 'ld.idx')
    .data.cols <- c('theta', 'theta.var', 'lodds', 'lfsr')

    .tab <- tab1 %>%
        select_(.dots = c(.cols, .data.cols)) %>%
            mutate(theta.se = sqrt(theta.var)) %>%
                select(-theta.var)

    ret <- tab2 %>%
        select_(.dots = c(.cols, .data.cols)) %>%            
            mutate(theta.se = sqrt(theta.var)) %>%
                select(-theta.var) %>%
                    left_join(.tab, by = .cols, suffix = c('.1', '.2')) %>%
                        na.omit()
}

stat.merged.tab <- merge.stat.tabs(tab1, tab2) %>%
    change.eqtl.name() %>%
    change.gwas.name()
```

# Results

## CaMMEL results

Full list of statistics generated on `20180602` are available:

* [All individuals](../result/20180602/obs/gene-pgc_ptsd_all_ea.txt.gz)
* [Military individuals](../result/20180602/obs/gene-pgc_ptsd_mil_ea.txt.gz)
* [Civilian individuals](../result/20180602/obs/gene-pgc_ptsd_civ_ea.txt.gz)

Full list of statistics generated on `20180605` are available:

* [All individuals](../result/20180605/obs/gene-pgc_ptsd_all_ea.txt.gz)
* [Military individuals](../result/20180605/obs/gene-pgc_ptsd_mil_ea.txt.gz)
* [Civilian individuals](../result/20180605/obs/gene-pgc_ptsd_civ_ea.txt.gz)

We used three sets of eQTL data to capture gene regulatory mechanisms
in the following relevant tissues.

1. DPLFC (dorsolateral prefrontal cortex) from postmortem samples of ROSMAP cohort. 

2. TCX (dorsolateral prefrontal cortex) from postmortem samples of Mayo cohort.

3. LCL (lymphoid cell line) from gEUVADIS study.

To avoid unwanted weak instrument bias, we tested mediation effects of
`r tab1 %>% select(hgnc, data) %>% unique() %>% nrow()` gene models of
`r tab1 %>% select(hgnc) %>% unique() %>% nrow()` unique genes after
stringent tests on gene expression heritability.  We retained genes if
and only if any of multivariate eQTL effects show significant non-zero
effects with posterior inclusion probability greater than 0.9.

## Concordance of mediation effects despite the infusion of random perturbation

```{r include = FALSE}

stat.high <- stat.merged.tab %>%
    filter(lfsr.1 < 0.1, lfsr.2 < 0.1)

Fig.rep <- gg.plot(stat.merged.tab, aes(x = lodds.1, y = lodds.2)) +
    xlab('Log odds in experiment 1') + ylab('Log odds in experiment 2') +
    geom_point(size = .5) +
    geom_point(data = stat.high, size = 2, col = 2, pch = 21) +
    facet_grid(gwas ~ data, scales = 'free')
```

Accuracy of causal mediation analysis depends on estimation of
unmediated effects.  We tested robustness of the mediation analysis
infusing artifactual mediators generated by permutation of eQTL
locations.  The following plots compare log-odds ratio of mediation
effect sizes between the first and second experiments.  Red circles
highlight `r nrow(stat.high)` genes strongly replicated in both
experiments.

```{r fig.width = 6, fig.height = 6}
print(Fig.rep)
```

```{r include = FALSE}
library(knitr)

temp <- stat.high %>%
    select(hgnc, data, gwas, starts_with('theta'), starts_with('lfsr')) %>%
    mutate(theta.1 = round(theta.1, 2) %&&% ' (' %&&% round(theta.se.1, 2) %&&% ')',
           theta.2 = round(theta.2, 2) %&&% ' (' %&&% round(theta.se.2, 2) %&&% ')') %>%
    select(-theta.se.1, -theta.se.2) %>%
    mutate(lfsr.1 = round(100 * lfsr.1), lfsr.2 = round(100 * lfsr.2)) %>%
    rename(tissue = data) %>%
    arrange(tissue, lfsr.1)
```

```{r echo = FALSE, results = 'asis'}
kable(temp %>% filter(gwas == 'All') %>% select(-gwas),
      caption = 'Significant genes in GWAS All',
      digits = 2)
```

```{r echo = FALSE, results = 'asis'}
kable(temp %>% filter(gwas == 'Mil') %>% select(-gwas),
      caption = 'Significant genes in GWAS Military',
      digits = 2)
```

```{r echo = FALSE, results = 'asis'}
kable(temp %>% filter(gwas == 'Civ') %>% select(-gwas),
      caption = 'Significant genes in GWAS Civilian',
      digits = 2)
```



